updater/osv/cvss.go

```diff
@@ -177,3 +194,136 @@ func fromCVSS(s string) (sev claircore.Severity, err error) {
 	}
 	return sev, nil
 }
+
+// FromCVSS2 is an attempt at an implementation of the formulas
+// described here: https://www.first.org/cvss/v2/guide
+func fromCVSS2(s string) (sev claircore.Severity, err error) {
+	ms := strings.Split(s, "/") // "m" as in "metric"
+	if len(ms) < 6 {
+		return 0, fmt.Errorf("bad vector: %q", s)
+	}
+	// Giant switch ahoy
+	var ns [6]float64
+	for _, m := range ms {
+		n, v, ok := strings.Cut(m, ":")
+		if !ok {
+			return 0, fmt.Errorf("bad metric: %q", m)
+		}
+		switch n {
+		// Base metrics:
+		case `AV`:
+			const i = 0
+			switch v {
+			case `N`:
+				ns[i] = 1
+			case `A`:
+				ns[i] = 0.646
+			case `L`:
+				ns[i] = 0.395
+			default:
+				return 0, fmt.Errorf("bad metric value: %q", m)
+			}
+		case `AC`:
+			const i = 1
+			switch v {
+			case `L`:
+				ns[i] = 0.71
+			case `M`:
+				ns[i] = 0.61
+			case `H`:
+				ns[i] = 0.35
+			default:
+				return 0, fmt.Errorf("bad metric value: %q", m)
+			}
+		case `Au`:
+			const i = 2
+			switch v {
+			case `M`:
+				ns[i] = 0.45
+			case `S`:
+				ns[i] = 0.56
+			case `N`:
+				ns[i] = 0.704
+			default:
+				return 0, fmt.Errorf("bad metric value: %q", m)
+			}
+		case `C`:
+			const i = 3
+			switch v {
+			case `C`:
+				ns[i] = 0.660
+			case `P`:
+				ns[i] = 0.275
+			case `N`:
+				ns[i] = 0
+			default:
+				return 0, fmt.Errorf("bad metric value: %q", m)
+			}
+		case `I`:
+			const i = 4
+			switch v {
+			case `C`:
+				ns[i] = 0.660
+			case `P`:
+				ns[i] = 0.275
+			case `N`:
+				ns[i] = 0
+			default:
+				return 0, fmt.Errorf("bad metric value: %q", m)
+			}
+		case `A`:
+			const i = 5
+			switch v {
+			case `C`:
+				ns[i] = 0.660
+			case `P`:
+				ns[i] = 0.275
+			case `N`:
+				ns[i] = 0
+			default:
+				return 0, fmt.Errorf("bad metric value: %q", m)
+			}
+		case `E`, `RL`, `RC`, `CDP`, `TD`, `CR`, `IR`, `AR`:
+			// Ignore temporal and environmental metrics.
+		default:
+			return 0, fmt.Errorf("bad metric: %q", m)
+		}
+	}
+
+	var score float64
+	exploitability := 20 * ns[ /*AV*/ 0] * ns[ /*AC*/ 1] * ns[ /*Au*/ 2]
+	impact := 10.41 * (1 - (1-ns[ /*C*/ 3])*(1-ns[ /*I*/ 4])*(1-ns[ /*A*/ 5]))
+	var fImpact float64
+	if impact != 0 {
+		fImpact = 1.176
+	}
+	score = ((0.6 * impact) + (0.4 * exploitability) - 1.5) * fImpact
+	// Reuse the CVSSv3 roundup function.
+	//
+	// The "real" score is truncated to one decimal position, but since we're
+	// only using it to calculate the normalized severity, I'm not going to worry
+	// about the precision.
+	i := int(score * 100_000)
+	if (i % 10_000) == 0 {
+		score = float64(i) / 100_000.0
+	} else {
+		score = ((float64(i) / 10_000) + 1) / 10.0
+	}
+	score = math.Min(score, 10)
+
+	switch {
```

Comment:
This is different for CVSSv2 according to https://nvd.nist.gov/vuln-metrics/cvss

0.0 <= x < 4.0 -> Low
4.0 <= x < 7.0 -> Medium
7.0 <= x <= 10.0 -> High